include/solvers/mf_navier_stokes_operators.h

```diff
@@ -0,0 +1,360 @@
+
+
+  /**
+   * @brief Interface to function in charge of computing the residual using a cell
+   * integral in a cell batch with gathering and scattering values. Should be
+   * overriden by derived classes.
+   *
+   * @param matrix_free Object that contains all data.
+   * @param dst Global vector where the final result is added.
+   * @param src Input vector with all values in all cells.
+   * @param range Range of the cell batch.
+   */
+  virtual void
+  local_evaluate_residual(
+    const MatrixFree<dim, number> &              matrix_free,
+    VectorType &                                 dst,
+    const VectorType &                           src,
+    const std::pair<unsigned int, unsigned int> &range) const;
+
+
+private:
+  /**
+   * @brief Get the refinement edges. Only needed if the local smoothing
+   * multigrid approach is used.
+   *
+   * @param matrix_free Object that contains all data.
+   * @return IndexSet Set containing the indices of the refinement edges.
+   */
+  static IndexSet
+  get_refinement_edges(const MatrixFree<dim, number> &matrix_free);
+
+protected:
+  MatrixFree<dim, number>                matrix_free;
+  AffineConstraints<number>              constraints;
+  mutable TrilinosWrappers::SparseMatrix system_matrix;
+  AlignedVector<VectorizedArray<number>> element_size;
+  unsigned int                           fe_degree;
+  const Function<dim> *                  forcing_function;
+  double                                 viscosity;
+
+  // Variables needed from the last newton step vector
+  Table<2, Tensor<1, dim + 1, VectorizedArray<number>>>
+    nonlinear_previous_values;
+  Table<2, Tensor<1, dim + 1, Tensor<1, dim, VectorizedArray<number>>>>
+    nonlinear_previous_gradient;
+  Table<2, Tensor<1, dim + 1, Tensor<1, dim, VectorizedArray<number>>>>
+    nonlinear_previous_hessian_diagonal;
+
+  // Variables needed for the local smoothing approach
+  std::vector<unsigned int>                      constrained_indices;
+  mutable std::vector<std::pair<number, number>> constrained_values;
+  std::vector<unsigned int>                      edge_constrained_indices;
+  bool has_edge_constrained_indices = false;
+  mutable std::vector<std::pair<number, number>> edge_constrained_values;
+  std::vector<bool>                              edge_constrained_cell;
+};
+
+/**
+ * @brief Class in charge of implementing the main function required to
+ * solve the Navier-Stokes equations using SUPG/PSPG stabilization and the
+ * matrix-free approach.
+ *
+ * @tparam dim An integer that denotes the number of spatial dimensions.
+ * @tparam number Abstract type for number across the class (i.e., double).
+ */
+template <int dim, typename number>
+class NavierStokesSUPGPSPGOperator
+  : public NavierStokesOperatorBase<dim, number>
+{
+public:
+  using FECellIntegrator = FEEvaluation<dim, -1, 0, dim + 1, number>;
+  using VectorType       = LinearAlgebra::distributed::Vector<number>;
```

Comment:
Does the using need to be remade for every inherited class?

Reply:
I think they cannot be inherited as it is right now. The only way would be to declare both classes inside a namespace and declare the using directive right before it.